recommenders/models/sasrec/sampler.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import numpy as np
+from multiprocessing import Process, Queue
+
+
+def random_neq(left, right, s):
+    t = np.random.randint(left, right)
+    while t in s:
+        t = np.random.randint(left, right)
+    return t
+
+
+def sample_function(
+    user_train, usernum, itemnum, batch_size, maxlen, result_queue, SEED
+):
+    """
+    Batch sampler that creates a sequence of negative items based on the
+    original sequence of items (positive) that the user has interacted with.
+    """
+
+    def sample():
+
+        user = np.random.randint(1, usernum + 1)
+        while len(user_train[user]) <= 1:
+            user = np.random.randint(1, usernum + 1)
+
+        seq = np.zeros([maxlen], dtype=np.int32)
+        pos = np.zeros([maxlen], dtype=np.int32)
+        neg = np.zeros([maxlen], dtype=np.int32)
+        nxt = user_train[user][-1]
+        idx = maxlen - 1
+
+        ts = set(user_train[user])
+        for i in reversed(user_train[user][:-1]):
+            seq[idx] = i
+            pos[idx] = nxt
+            if nxt != 0:
+                neg[idx] = random_neq(1, itemnum + 1, ts)
+            nxt = i
+            idx -= 1
+            if idx == -1:
+                break
+
+        return (user, seq, pos, neg)
+
+    np.random.seed(SEED)
+    while True:
+        one_batch = []
+        for i in range(batch_size):
+            one_batch.append(sample())
+
+        result_queue.put(zip(*one_batch))
+
+
+class WarpSampler(object):
+    def __init__(self, User, usernum, itemnum, batch_size=64, maxlen=10, n_workers=1):
+        self.result_queue = Queue(maxsize=n_workers * 10)
+        self.processors = []
+        for i in range(n_workers):
+            self.processors.append(
+                Process(
+                    target=sample_function,
+                    args=(
+                        User,
+                        usernum,
+                        itemnum,
+                        batch_size,
+                        maxlen,
+                        self.result_queue,
+                        np.random.randint(2e9),
+                    ),
+                )
+            )
+            self.processors[-1].daemon = True
+            self.processors[-1].start()
+
```

Comment:
this is super nice Abir!
Do you have a sense on what is the speed up you got with this multiprocess sampler in comparison with a normal one?